File: backend/models/recipe_share.py
"""Public recipe sharing models — share links + clone."""

from datetime import datetime
from typing import Any, Dict, Optional

from pydantic import BaseModel, Field


class ShareLink(BaseModel):
    recipe_id: str
    slug: str
    url: str                    # full shareable URL constructed by service
    view_count: int = 0
    save_count: int = 0
    created_at: datetime
    is_public: bool = True


class CreateShareLinkResponse(BaseModel):
    link: ShareLink


class PublicRecipeView(BaseModel):
    """Sanitized public payload — no owner PII."""
    slug: str
    name: str
    description: Optional[str] = None
    image_url: Optional[str] = None
    servings: int
    prep_time_minutes: Optional[int] = None
    cook_time_minutes: Optional[int] = None
    instructions: Optional[str] = None
    category: Optional[str] = None
    cuisine: Optional[str] = None
    tags: list = Field(default_factory=list)
    cooking_method: Optional[str] = None
    cooked_yield_grams: Optional[float] = None
    calories_per_serving: Optional[int] = None
    protein_per_serving_g: Optional[float] = None
    carbs_per_serving_g: Optional[float] = None
    fat_per_serving_g: Optional[float] = None
    fiber_per_serving_g: Optional[float] = None
    micronutrients_per_serving: Optional[Dict[str, Any]] = None
    ingredients: list = Field(default_factory=list)
    times_logged: int = 0
    view_count: int = 0
    save_count: int = 0
    author_display_name: Optional[str] = None  # falls back to "Anonymous chef"


class CloneRecipeResponse(BaseModel):
    new_recipe_id: str
    already_saved: bool = False
    message: str

File: mobile/flutter/lib/data/models/recipe_share.dart
/// Public recipe sharing models.
library;

class ShareLink {
  final String recipeId;
  final String slug;
  final String url;
  final int viewCount;
  final int saveCount;
  final DateTime createdAt;
  final bool isPublic;

  const ShareLink({
    required this.recipeId,
    required this.slug,
    required this.url,
    required this.createdAt,
    this.viewCount = 0,
    this.saveCount = 0,
    this.isPublic = true,
  });

  factory ShareLink.fromJson(Map<String, dynamic> j) => ShareLink(
        recipeId: j['recipe_id'] as String,
        slug: j['slug'] as String,
        url: j['url'] as String,
        viewCount: j['view_count'] as int? ?? 0,
        saveCount: j['save_count'] as int? ?? 0,
        createdAt: DateTime.parse(j['created_at'] as String),
        isPublic: j['is_public'] as bool? ?? true,
      );
}

class PublicRecipeView {
  final String slug;
  final String name;
  final String? description;
  final String? imageUrl;
  final int servings;
  final int? prepTimeMinutes;
  final int? cookTimeMinutes;
  final String? instructions;
  final String? category;
  final String? cuisine;
  final List<String> tags;
  final String? cookingMethod;
  final double? cookedYieldGrams;
  final int? caloriesPerServing;
  final double? proteinPerServingG;
  final double? carbsPerServingG;
  final double? fatPerServingG;
  final double? fiberPerServingG;
  final List<Map<String, dynamic>> ingredients;
  final int timesLogged;
  final int viewCount;
  final int saveCount;
  final String? authorDisplayName;

  const PublicRecipeView({
    required this.slug,
    required this.name,
    required this.servings,
    required this.ingredients,
    required this.timesLogged,
    required this.viewCount,
    required this.saveCount,
    this.description,
    this.imageUrl,
    this.prepTimeMinutes,
    this.cookTimeMinutes,
    this.instructions,
    this.category,
    this.cuisine,
    this.tags = const [],
    this.cookingMethod,
    this.cookedYieldGrams,
    this.caloriesPerServing,
    this.proteinPerServingG,
    this.carbsPerServingG,
    this.fatPerServingG,
    this.fiberPerServingG,
    this.authorDisplayName,
  });

  factory PublicRecipeView.fromJson(Map<String, dynamic> j) => PublicRecipeView(
        slug: j['slug'] as String,
        name: j['name'] as String,
        description: j['description'] as String?,
        imageUrl: j['image_url'] as String?,
        servings: j['servings'] as int? ?? 1,
        prepTimeMinutes: j['prep_time_minutes'] as int?,
        cookTimeMinutes: j['cook_time_minutes'] as int?,
        instructions: j['instructions'] as String?,
        category: j['category'] as String?,
        cuisine: j['cuisine'] as String?,
        tags: (j['tags'] as List? ?? []).map((e) => e as String).toList(),
        cookingMethod: j['cooking_method'] as String?,
        cookedYieldGrams: (j['cooked_yield_grams'] as num?)?.toDouble(),
        caloriesPerServing: j['calories_per_serving'] as int?,
        proteinPerServingG: (j['protein_per_serving_g'] as num?)?.toDouble(),
        carbsPerServingG: (j['carbs_per_serving_g'] as num?)?.toDouble(),
        fatPerServingG: (j['fat_per_serving_g'] as num?)?.toDouble(),
        fiberPerServingG: (j['fiber_per_serving_g'] as num?)?.toDouble(),
        ingredients: (j['ingredients'] as List? ?? []).cast<Map<String, dynamic>>(),
        timesLogged: j['times_logged'] as int? ?? 0,
        viewCount: j['view_count'] as int? ?? 0,
        saveCount: j['save_count'] as int? ?? 0,
        authorDisplayName: j['author_display_name'] as String?,
      );
}

class CloneRecipeResponse {
  final String newRecipeId;
  final bool alreadySaved;
  final String message;

  const CloneRecipeResponse({
    required this.newRecipeId,
    required this.message,
    this.alreadySaved = false,
  });

  factory CloneRecipeResponse.fromJson(Map<String, dynamic> j) => CloneRecipeResponse(
        newRecipeId: j['new_recipe_id'] as String,
        alreadySaved: j['already_saved'] as bool? ?? false,
        message: (j['message'] ?? '') as String,
      );
}
